File: programmers/lv1/숫자_짝꿍.js
// https://school.programmers.co.kr/learn/courses/30/lessons/131128

function solution(X, Y) {
  let answer = '';
  const xNum = {}, yNum = {};
  [...X].forEach(x => {
    if(x in xNum)   xNum[x]++;
    else xNum[x] = 1;
  });
  [...Y].forEach(y => {
    if(y in yNum)   yNum[y]++;
    else yNum[y] = 1;
  });
  for(const x in xNum){
    for(const y in yNum){
      if(x === y) answer += x.toString().repeat(Math.min(xNum[x], yNum[y]));
    }
  }
  if(answer === "")   answer = '-1';
  if(Number(answer) === 0)    answer = '0';
  return [...answer].sort((a, b) => b - a).join('');
}

File: programmers/lv1/숫자_짝꿍.py
from collections import Counter
def solution(X, Y):
    answer = ''
    cntX = Counter(X)
    cntY = Counter(Y)

    for i in range(9, -1, -1):
        answer += str(i) * min(cntX[str(i)], cntY[str(i)])
    if answer == '':
        return '-1'
    elif answer[0] == '0':
        return '0'
    return answer
# 공통된 숫자가 0으로만 구성되어 있을 때 answer[0]이 0인지 확인한다.
# int(answer) == 0 으로 확인했을 때 시간초과 발생함
